File: 2-trap/trap.f90
MODULE precision
  INTEGER ,PARAMETER:: long=SELECTED_REAL_KIND(15, 307) !(6,37) for real(float), (15,307) for double
END MODULE precision

module time_ftcs
    contains
      function timestamp() result(str)
        implicit none
        character(len=20)                     :: str
        integer                               :: values(8)
        character(len=4)                      :: year
        character(len=2)                      :: month
        character(len=2)                      :: day, hour, minute, second
        character(len=5)                      :: zone
    
        ! Get current time
        call date_and_time(VALUES=values, ZONE=zone)  
        write(year,'(i4.4)')    values(1)
        write(month,'(i2.2)')   values(2)
        write(day,'(i2.2)')     values(3)
        write(hour,'(i2.2)')    values(5)
        write(minute,'(i2.2)')  values(6)
        write(second,'(i2.2)')  values(7)
    
        str = day//'-'//hour//'-'//minute//'-'//second
      end function timestamp
end module

program main
    use precision
    use time_ftcs
    implicit none

    INTERFACE 
    
    SUBROUTINE Trap(a, b, n, h, integral)
        use precision
        implicit none
        REAL(long), INTENT(IN) :: a
        REAL(long), INTENT(IN) :: b
        INTEGER, INTENT(IN) :: n
        REAL(long), INTENT(IN) :: h
        REAL(long), INTENT(OUT) :: integral
    END SUBROUTINE Trap

    subroutine OpenFile(a, b, n)
        use precision
        implicit none
        REAL(long), INTENT(OUT) :: a
        REAL(long), INTENT(OUT) :: b
        INTEGER, INTENT(OUT) :: n
    end subroutine OpenFile

    END INTERFACE

    ! Declaration of variables
    REAL(long) :: integral
    REAL(long) :: a
    REAL(long) :: b
    REAL(long) :: h
    INTEGER :: n
    REAL(long) :: estimation                            ! Estimacion del calculo
    REAL(long) :: error                                 ! Error encontrado
    REAL(long) :: elapsed                               ! Tiempo que tomo ejecutarse el programa
    REAL(long) :: start_clock                           ! Tiempo de inicio
    REAL(long) :: stop_clock                            ! Tiempo de termino
    CHARACTER(len=38) :: start                          ! Hora de inicio
    CHARACTER(len=38) :: end                            ! Hora de termino

    ! Establece la hora de inicio
    start = timestamp()
    call CPU_TIME(start_clock)

    ! Assignment
    call OpenFile(a, b, n)

    h = (b - a) / n;
    call Trap (a, b, n, h, integral)

    ! Estimacion del resultado
    estimation = 9.0_long

    ! Calcula el % de error
    error = DABS(integral - estimation) / estimation * 100.0_long

    ! Establece la hora en que termino de calcular
    end = timestamp()
    call CPU_TIME(stop_clock)

    ! Calcula el tiempo que tomo ejecutar el programa
    elapsed =  stop_clock - start_clock

    ! Guarda el resultado en un archivo
    call SaveFile(start, end, error, elapsed, N)

end program main

! ------------------------------------------------------------------
! Function:    Trap
! Purpose:     Estimate integral from a to b of f using trap rule and
!              n trapezoids
! Input args:  a, b, n, h
! Return val:  Estimate of the integral 
! ------------------------------------------------------------------
SUBROUTINE Trap(a, b, n, h, integral)
    use precision
    implicit none
    REAL(long), INTENT(IN) :: a
    REAL(long), INTENT(IN) :: b
    INTEGER, INTENT(IN) :: n
    REAL(long), INTENT(IN) :: h
    REAL(long), INTENT(OUT) :: integral
    INTEGER :: k

    integral = ( a**2 + b**2 ) / 2_long

    do k = 1, n - 1, 1
        integral = integral + (a + k * h) ** 2
    end do

    integral = integral * h

END SUBROUTINE Trap

subroutine OpenFile(a, b, n)
    use precision
    implicit none
    REAL(long), INTENT(OUT) :: a
    REAL(long), INTENT(OUT) :: b
    INTEGER, INTENT(OUT) :: n
    INTEGER :: stat
    CHARACTER(len=30) :: input_file

    input_file = 'parameters.dat'

    OPEN(UNIT=1,file=input_file,ACTION="READ",IOSTAT=stat,STATUS='OLD')
    READ(1, *) a
    READ(1, *) b
    READ(1,*) n
 
    CLOSE(1)

end subroutine OpenFile

! Crea un archivo de salida con la hora de inicio, de termino y el tiempo que tomo correr el programa
! Asi como el porcentaje de error
subroutine SaveFile(start, end, error, elapsed, N)
    use precision
    CHARACTER(len=38), INTENT(IN) :: start
    CHARACTER(len=38), INTENT(IN) :: end
    REAL(long), INTENT(IN) :: error
    REAL(long), INTENT(IN) :: elapsed
    INTEGER, INTENT(IN) :: N
    CHARACTER(len = 54) :: file_name
    CHARACTER(len = 8) :: Nstring

    write(Nstring, "(I4)") N 

    file_name = 'serial-f90-'//Nstring//'-'//trim(end)//'.txt'
    
    open (unit=10,file=file_name)
    write(10,*) 'Hora de inicio'
    write(10,*) start
    write(10,*) 'Hora de termino'
    write(10,*) end
    write(10,*) 'Tiempo de ejecucion'
    write(10,*) elapsed
    write(10,*) 'Error'
    write(10,*) error
    close(10)

end subroutine SaveFile

File: 2-trap/trap.c
/* File:    trap.c
 * Purpose: Calculate definite integral using trapezoidal 
 *          rule.
 *
 * Input:   a, b, n
 * Output:  Estimate of integral from a to b of f(x)
 *          using n trapezoids.
 *
 * Compile: gcc -g -Wall -o trap trap.c
 * Usage:   ./trap
 *
 * Note:    The function f(x) is hardwired.
 *
 * IPP:     Section 3.2.1 (pp. 94 and ff.) and 5.2 (p. 216)
 */

#include <stdio.h>
#include <math.h>
#include <time.h>

double Trap(double a, double b, long n, double h);
void OpenFile(double *a, double *b, long *n);
void SaveFile(struct tm *start, struct tm *end, double error, double elapsed, long N);

int main(int argc, char *argv[])
{
    double integral;     /* Store result in integral */
    double a;            /* Left endpoints */
    double b;            /* Right endpoints */
    double h;            /* Height of trapezoids */
    long n;              /* Number of trapezoids */
    double estimation;   // Estimacion del calculo
    double error;        // Error encontrado
    double elapsed;      // Tiempo que tomo ejecutarse el programa
    time_t t;            // Variable de tiempo
    struct tm *start;    // Hora de inicio
    struct tm *end;      // Hora de termino
    clock_t start_clock; // Tiempo de inicio
    clock_t stop_clock;  // Tiempo de termino

    // Inicia la variable de tiempo
    t = time(NULL);

    // Establece la hora de inicio
    start = localtime(&t);
    start_clock = clock();

    //
    OpenFile(&a, &b, &n);

    h = (b - a) / n;
    integral = Trap(a, b, n, h);

    // Estimacion del resultado
    estimation = 9.0;

    // Calcula el porcentaje de error.
    error = fabs(integral - estimation) / estimation * 100.0;

    // Establece la hora en que termino de calcular
    end = localtime(&t);
    stop_clock = clock();

    // Calcula el tiempo que tomo ejecutar el programa
    elapsed = (double)(stop_clock - start_clock) / CLOCKS_PER_SEC;

    // Guarda el resultado en un archivo
    SaveFile(start, end, error, elapsed, n);

    return 0;
}

/*------------------------------------------------------------------
 * Function:    Trap
 * Purpose:     Estimate integral from a to b of f using trap rule and
 *              n trapezoids
 * Input args:  a, b, n, h
 * Return val:  Estimate of the integral 
 */
double Trap(double a, double b, long n, double h)
{
    double integral;
    long k;

    integral = (a * a + b * b) / 2.0;

    for (k = 1; k <= n - 1; k++)
        integral += (a + k * h) * (a + k * h);

    integral = integral * h;

    return integral;
}

void OpenFile(double *a, double *b, long *n)
{
    FILE *file;
    char *input_file;

    input_file = "parameters.dat";

    file = fopen(input_file, "r");

    if (file == NULL)
    {
        printf("No se encontro el archivo \"parameters.dat\" se usaran parametros por defecto.\n");
        *a = 0.0;
        *b = 3.0;
        *n = 1024;
    }
    else
        fscanf(file, "%lf %lf %ld", a, b, n);

    fclose(file);
}

// Crea un archivo de salida con la hora de inicio, de termino y el tiempo que tomo correr el programa
// Asi como el porcentaje de error
void SaveFile(
    struct tm *start, // Hora de inicio
    struct tm *end,   // Hora de termino
    double error,     // Porcentaje de error
    double elapsed,   // Tiempo que paso
    long N            // Dimension de la matriz
)
{
    FILE *file;
    char file_name[64];
    char output[50];

    sprintf(file_name, "serial-c-%ld-%d-%d-%d-%d.txt", N, end->tm_mday, end->tm_hour, end->tm_min, end->tm_sec);

    file = fopen(file_name, "w+");

    strftime(output, sizeof(output), "%c", start);
    fprintf(file, "Hora de inicio\n%s\n", output);

    strftime(output, sizeof(output), "%c", end);
    fprintf(file, "Hora de termino\n%s\n", output);

    fprintf(file, "Tiempo de ejecucion\n%.15lf\n", elapsed);

    fprintf(file, "Error\n%.15le\n", error);

    fclose(file);
}